### PR TITLE
Get rid of lambda within `AST::TypePath` and provide a method to return a reference

### PR DESCRIPTION
**Pull request by diohabara**
## Related issue

This PR will fix <https://github.com/Rust-GCC/gccrs/issues/718>

- \[x] GCC development requires copyright assignment or the Developer's Certificate of Origin sign-off, see https://gcc.gnu.org/contribute.html or https://gcc.gnu.org/dco.html
- \[x] Read contributing guidlines
- \[x] `make check-rust` passes locally
- \[x] Run `clang-format`
- \[x] Added any relevant test cases to `gcc/testsuite/rust/`

Note that you can skip the above if you are just opening a WIP PR in
order to get feedback.

### CODE REVIEW

File: gcc/rust/hir/rust-ast-lower-type.h [hunk 1]
```diff
@@ -62,20 +62,18 @@ class ASTLowerTypePath : public ASTLoweringBase
...
-	std::unique_ptr<HIR::TypePathSegment> (translated_segment));
-      return true;
-    });
+    for (auto &seg : path.get_segments ())
+      {
+	translated_segment = nullptr;
+	seg->accept_vis (*this);
+	if (translated_segment == nullptr)
+	  {
+	    rust_fatal_error (seg->get_locus (),
```
Comment (philberty) [inline gcc/rust/hir/rust-ast-lower-type.h:71]: This is not related to your PR but lets raise an issue about this fatal_error it should be an assertion

File: gcc/rust/hir/rust-ast-lower.cc [hunk 2]
```diff
@@ -490,20 +490,18 @@ ASTLowerQualifiedPathInType::visit (AST::QualifiedPathInType &path)
...
-      std::unique_ptr<HIR::TypePathSegment> (translated_segment));
-    return true;
-  });
+  for (auto &seg : path.get_segments ())
+    {
+      translated_segment = nullptr;
+      seg->accept_vis (*this);
+      if (translated_segment == nullptr)
+	{
+	  rust_fatal_error (seg->get_locus (),
```
Comment (philberty) [inline gcc/rust/hir/rust-ast-lower.cc:499]: Same here

**Review by philberty — CHANGES_REQUESTED** [review]
LGTM you just need to squash these fixup commits:

$ git rebase -i HEAD~3

Is what i use to do it.

File: gcc/rust/hir/rust-ast-lower-type.h (hunk 1 above)
Comment (diohabara) [reply gcc/rust/hir/rust-ast-lower-type.h:71]: Thanks for you quick review!

Are you saying I should replace `rust_fatal_error` with another function or another procedure?

Comment (philberty) [reply gcc/rust/hir/rust-ast-lower-type.h:71]: Not as part of this pr. I raised a separate issue for that.

**Review by tschwinge — CHANGES_REQUESTED** [review]
Please again force-push after updating the commit log (`git commit --amend`) into standard format (see `git log`), and then we're good to go.

**Review by philberty — APPROVED** [review]
LGTM thanks again! :D nice job

**Pull request merged**